src/utility/svd.jl

```diff
@@ -0,0 +1,303 @@
+The number of targeted singular values is set via the `TruncationSpace` in `ProjectorAlg`.
+In particular, this make it possible to specify the targeted singular values block-wise.
+In case the symmetry block is too small as compared to the number of singular values, or
+the iterative SVD didn't converge, the algorithm falls back to a dense SVD.
+"""
+@kwdef struct IterSVD
+    alg::KrylovKit.GKL = KrylovKit.GKL(; tol=1e-14, krylovdim=25)
+    fallback_threshold::Float64 = Inf
+end
+
+# Compute SVD data block-wise using KrylovKit algorithm
+function TensorKit._tsvd!(
+    t, alg::IterSVD, trunc::Union{NoTruncation,TruncationSpace}, p::Real=2
+)
+    # early return
+    if isempty(blocksectors(t))
+        truncerr = zero(real(scalartype(t)))
+        return _empty_svdtensors(t)..., truncerr
+    end
+
+    Udata, Σdata, Vdata, dims = _compute_svddata!(t, alg, trunc)
+    U, S, V = _create_svdtensors(t, Udata, Σdata, Vdata, spacetype(t)(dims))
+    truncerr = trunc isa NoTruncation ? abs(zero(scalartype(t))) : norm(U * S * V - t, p)
+
+    return U, S, V, truncerr
+end
+function TensorKit._compute_svddata!(
+    t::TensorMap, alg::IterSVD, trunc::Union{NoTruncation,TruncationSpace}
+)
+    InnerProductStyle(t) === EuclideanProduct() || throw_invalid_innerproduct(:tsvd!)
+    I = sectortype(t)
+    A = storagetype(t)
+    Udata = SectorDict{I,A}()
+    Vdata = SectorDict{I,A}()
+    dims = SectorDict{I,Int}()
+    local Sdata
+    for (c, b) in blocks(t)
+        x₀ = randn(eltype(b), size(b, 1))
+        howmany = trunc isa NoTruncation ? minimum(size(b)) : blockdim(trunc.space, c)
+
+        if howmany / minimum(size(b)) > alg.fallback_threshold  # Use dense SVD for small blocks
+            U, S, V = TensorKit.MatrixAlgebra.svd!(b, TensorKit.SVD())
+            Udata[c] = U[:, 1:howmany]
+            Vdata[c] = V[1:howmany, :]
+        else
+            S, lvecs, rvecs, info = KrylovKit.svdsolve(b, x₀, howmany, :LR, alg.alg)
+            if info.converged < howmany  # Fall back to dense SVD if not properly converged
+                U, S, V = TensorKit.MatrixAlgebra.svd!(b, TensorKit.SVD())
+                Udata[c] = U[:, 1:howmany]
+                Vdata[c] = V[1:howmany, :]
+            else  # Slice in case more values were converged than requested
+                Udata[c] = stack(view(lvecs, 1:howmany))
+                Vdata[c] = stack(conj, view(rvecs, 1:howmany); dims=1)
+            end
+        end
+
+        S = @view S[1:howmany]
+        if @isdefined Sdata # cannot easily infer the type of Σ, so use this construction
+            Sdata[c] = S
+        else
+            Sdata = SectorDict(c => S)
+        end
+        dims[c] = length(S)
+    end
+    return Udata, Sdata, Vdata, dims
+end
+
+"""
+    struct CompleteSVDAdjoint(; lorentz_broadening = 0.0)
+
+Wrapper around the complete `TensorKit.tsvd!` rrule which requires computing the full SVD.
+"""
+@kwdef struct CompleteSVDAdjoint
+    lorentz_broadening::Float64 = 0.0
```

Comment:
But that is already contained in the `SVDrrule` wrapper, right? Then it shouldn't be needed inside the adjoint structs.

Reply:
Aha, I see, I misunderstood, somehow I thought this was the SVDrrule wrapper already. It seems like now, there are quite a lot of these structs, can we converge on something like:
```julia
struct SVDAdjoint{A,B,E}
    alg_fwd::A
    alg_rrule::B
    broadening::E
end
```
where `alg_rrule=nothing` for `alg_fwd= SVD()` or `SDD()`, and otherwise this is just the algorithm?

This probably still requires some dummy struct for dispatching with the "old/wrong" rrule, but otherwise it might be just a little too convoluted

Reply:
Yes I can see that now the struct situation is too convoluted, so I will try to cut down on that by using `SVDAdjoint`. Let me see, if I can clean it up...